File: src/global_optimization/Cells/Sphere.py
from __future__ import annotations
from math import sqrt
from skimage.draw import disk, circle_perimeter_aa
from typing import DefaultDict

from .mathhelper import Vector

from .Cell import Cell, CellParams, PerturbParams, CellConfig


class SphereConfig(CellConfig):
    x: PerturbParams
    y: PerturbParams
    z: PerturbParams
    radius: PerturbParams
    minRadius: float
    maxRadius: float

class SphereParams(CellParams):
    x: float
    y: float
    z: float = 0.0
    radius: float
    # opacity: float = 1.0
    # split_alpha: float = 0.0


class Sphere(Cell):
    """The Sphere class represents a spherical bacterium."""
    paramClass = SphereParams
    cellConfig: SphereConfig

    def __init__(self, init_props: SphereParams):
        # destructure the properties of the sphere
        name = init_props.name
        x = init_props.x
        y = init_props.y
        z = init_props.z
        radius = init_props.radius
        # opacity = init_props.opacity
        # split_alpha = init_props.split_alpha

        self._name = name
        self._position = Vector([x, y, z])
        self._radius = radius
        self._rotation = 0
        # self._opacity = opacity
        # self._split_alpha = split_alpha
        self.dormant = False

    def draw(self, image, simulation_config, cell_map = None, z = 0):
        """Draws the cell by adding the given value to the image."""
        if self.dormant:
            return

        current_radius = self.get_radius_at(z)
        if current_radius <= 0:
            return

        background_color = simulation_config.background_color
        cell_color = simulation_config.cell_color

        rr, cc = disk((self._position.y, self._position.x), current_radius, shape=image.shape)
        image[rr, cc] = cell_color


    def draw_outline(self, image, color, z = 0):
        """Draws the outline of the cell over a color image."""
        if self.dormant:
            return

        current_radius = self.get_radius_at(z)
        if current_radius <= 0:
            return

        rr, cc, val = circle_perimeter_aa(round(self._position.y), round(self._position.x), round(current_radius), shape=image.shape)
        image[rr, cc] = color

    def split(self, alpha):
        """Splits a cell into two cells with a ratio determined by alpha."""
        pass
        # if self._needs_refresh:
        #     self._refresh()
        #
        # direction = Vector([cos(self._rotation), sin(self._rotation), 0])
        # unit = self._length*direction
        #
        # front = self._position + unit/2
        # back = self._position - unit/2
        # center = self._position + (0.5 - float(alpha))*unit
        #
        # position1 = (front + center)/2
        # position2 = (center + back)/2
        #
        # cell1 = Sphere(
        #     self._name + '0',
        #     position1.x, position1.y,
        #     self._width, self._length*float(alpha),
        #     self._rotation, alpha, self._opacity)
        #
        # cell2 = Sphere(
        #     self._name + '1',
        #     position2.x, position2.y,
        #     self._width, self._length*(1 - float(alpha)),
        #     self._rotation, alpha, self._opacity)
        #
        # return cell1, cell2

    def combine(self, cell):
        """Combines this cell with another cell."""
        pass
        # if self._needs_refresh:
        #     self._refresh()
        #
        # if cell._needs_refresh:
        #     cell._refresh()
        #
        # separation = self._position - cell._position
        # direction = separation/sqrt(separation@separation)
        #
        # # get combined front
        # direction1 = Vector([cos(self._rotation), sin(self._rotation), 0])
        # distance1 = self._length - self._width
        # if direction1@direction >= 0:
        #     head1 = self._position + distance1*direction1/2
        # else:
        #     head1 = self._position - distance1*direction1/2
        # extent1 = head1 + self._width*direction/2
        # front = self._position + ((extent1 - self._position)@direction)*direction
        #
        # # get combined back
        # direction2 = Vector([cos(cell._rotation), sin(cell._rotation), 0])
        # distance2 = cell._length - cell._width
        # if direction2@direction >= 0:
        #     tail2 = cell._position - distance2*direction2/2
        # else:
        #     tail2 = cell._position + distance2*direction2/2
        # extent2 = tail2 - cell._width*direction/2
        # back = cell._position + ((extent2 - cell._position)@direction)*direction
        #
        # # create new cell
        # position = (front + back)/2
        # rotation = atan2(direction.y, direction.x)
        # width = (self._width + cell._width)/2
        # length = sqrt((front - back)@(front - back))
        #
        # return Sphere(
        #     self._name[:-1],
        #     position.x, position.y,
        #     width, length,
        #     rotation, "combined alpha unknown", (self._opacity + cell.opacity)/2)

    def get_perturbed_cell(self):
        return Sphere(SphereParams(
            name=self._name,
            x=self._position.x + Sphere.cellConfig.x.get_perturb_offset(),
            y=self._position.y + Sphere.cellConfig.y.get_perturb_offset(),
            z=self._position.z + Sphere.cellConfig.z.get_perturb_offset(),
            radius=self._radius + Sphere.cellConfig.radius.get_perturb_offset(),
        ))
    
    def get_paramaterized_cell(self, params: DefaultDict[str, float]):
        # if no params, default to using perterb from files (used during simulated annealing)
        return Sphere(SphereParams(
            name=self._name,
            x=self._position.x + params['x'],
            y=self._position.y + params['y'],
            z=self._position.z + params['z'],
            radius=min(max(Sphere.cellConfig.minRadius, self._radius + params['radius']), Sphere.cellConfig.maxRadius),
        ))

    def get_radius_at(self, z: float):
        """Returns the radius of the sphere at a given z value."""
        if abs(self._position.z - z) > self._radius:
            return 0
        return sqrt((self._radius) ** 2 - (self._position.z - z) ** 2)

    def __repr__(self):
        return (f'Sphere('
                f'name="{self._name}", '
                f'x={self._position.x:.2f}, y={self._position.y:.2f}, z={self._position.z:.2f}, '
                f'radius={self._radius:.2f}')

    def get_cell_params(self):
        return SphereParams(
            name=self._name,
            x=self._position.x,
            y=self._position.y,
            z=self._position.z,
            radius=self._radius,
            # opacity=self._opacity,
            # split_alpha=self._split_alpha
        )
    
    def calculate_corners(self):
        """
        Calculate the minimum and maximum corners of the sphere.

        :return: A tuple of two lists, each of size 3. The first list represents the minimum corner 
                (x, y, z), and the second list represents the maximum corner (x, y, z).
        """
        min_corner = [self._position.x - self._radius, self._position.y - self._radius, self._position.z - self._radius]
        max_corner = [self._position.x + self._radius, self._position.y + self._radius, self._position.z + self._radius]
        return min_corner, max_corner

    def calculate_minimum_box(self, perturbed_cell: Sphere):
        """
        Calculate the minimum box that could encompass the current sphere and a given sphere.

        :param perturbed_cell: The second sphere that defines the dimensions of the box.
        :return: A tuple of two lists, each of size 3. The first list represents the minimum corner 
                (x, y, z) of the box, and the second list represents the maximum corner (x, y, z) of the box.
        """
        cell1_min_corner, cell1_max_corner = self.calculate_corners()
        cell2_min_corner, cell2_max_corner = perturbed_cell.calculate_corners()

        # Find the minimum and maximum coordinates among the spheres' corners
        min_corner = [min(cell1_min_corner[i], cell2_min_corner[i]) for i in range(3)]
        max_corner = [max(cell1_max_corner[i], cell2_max_corner[i]) for i in range(3)]
            
        # Return the minimum and maximum corners of the box
        return min_corner, max_corner
        

File: 3d/C++/src/Cells/Sphere.cpp
#include "Sphere.hpp"

SphereParams Sphere::paramClass = SphereParams();
SphereConfig Sphere::cellConfig = SphereConfig();



double Sphere::get_radius_at(double z) const {
    if (std::abs(_position.z - z) > _radius) {
        return 0;
    }
    return std::sqrt((_radius * _radius) - ((_position.z - z) * (_position.z - z)));
}


void Sphere::draw(cv::Mat & image, SimulationConfig simulationConfig, cv::Mat * cellMap, float z) const  {
    if (dormant) {
        return;
    }

    float currentRadius = get_radius_at(z);
    if (currentRadius <= 0) {
        return;
    }

    float background_color = simulationConfig.background_color;
    float cell_color = simulationConfig.cell_color;

    cv::Point center(_position.x, _position.y);
    cv::circle(image, center, static_cast<int>(currentRadius), cv::Scalar(cell_color), -1);
}

void Sphere::draw_outline(cv::Mat& image, float color, float z) const {
    if (dormant) {
        return;
    }
    float outlineColor = color;
    float currentRadius = get_radius_at(z);
    if (currentRadius <= 0) {
        return;
    }
    cv::Point center(_position.x, _position.y);
    int thickness = 1;
    cv::circle(image, center, static_cast<int>(round(currentRadius)), cv::Scalar(outlineColor), thickness, cv::LINE_AA);
}

Cell* Sphere::get_perturbed_cell() const {
    SphereParams sphereParams(
            _name,
            _position.x + cellConfig.x.get_perturb_offset(),
            _position.y + cellConfig.y.get_perturb_offset(),
            _position.z + cellConfig.z.get_perturb_offset(),
            _radius + cellConfig.radius.get_perturb_offset()
    );
    return new Sphere(sphereParams);
}

Cell* Sphere::get_parameterized_cell(std::unordered_map<std::string, float> params) const  {
    float xOffset = params["x"];
    float yOffset = params["y"];
    float zOffset = params["z"];
    float radiusOffset = params["radius"];

    if (params.empty()) {
        xOffset = Sphere::cellConfig.x.get_perturb_offset();
        yOffset = Sphere::cellConfig.y.get_perturb_offset();
        zOffset = Sphere::cellConfig.z.get_perturb_offset();
        radiusOffset = Sphere::cellConfig.radius.get_perturb_offset();
    }

    float newRadius = fmin(fmax(Sphere::cellConfig.minRadius, _radius + radiusOffset), Sphere::cellConfig.maxRadius);
    SphereParams sphereParams(
            _name,
            _position.x + xOffset,
            _position.y + yOffset,
            _position.z + zOffset,
            newRadius
    );
    return new Sphere(sphereParams);
}

std::tuple<Cell*, Cell*, bool> Sphere::get_split_cells() const  {
    double theta = ((double)rand() / RAND_MAX) * 2 * M_PI;
    double phi = ((double)rand() / RAND_MAX) * M_PI;

    cv::Point3f split_axis(
            sin(phi) * cos(theta),
            sin(phi) * sin(theta),
            cos(phi));

    cv::Point3f offset = split_axis * (_radius / 2.0);
    cv::Point3f new_position1 = _position + offset;
    cv::Point3f new_position2 = _position - offset;

    double halfRadius = _radius / 2.0;

    Sphere cell1(SphereParams(_name + "0", new_position1.x, new_position1.y, new_position1.z, halfRadius));
    Sphere cell2(SphereParams(_name + "1", new_position2.x, new_position2.y, new_position2.z, halfRadius));

    bool constraints = cell1.check_constraints() && cell2.check_constraints();

    //Cell cell1Base = static_cast<Cell>(cell1);
    //Cell cell2Base = static_cast<Cell>(cell2);//convert Sphere to Cell

    return std::make_tuple(new Sphere(cell1), new Sphere(cell2), constraints);
}

bool Sphere::check_constraints() const {
    SphereConfig config;
    return (config.minRadius <= _radius) && (_radius <= config.maxRadius);
}

float Sphere::get_radius_at(float z) {
    if (std::abs(_position.z - z) > _radius) {
        return 0;
    }
    return std::sqrt((_radius * _radius) - ((_position.z - z) * (_position.z - z)));
}

CellParams Sphere::get_cell_params() const  {
    return SphereParams(_name, _position.x, _position.y, _position.z, _radius);
}

std::pair<std::vector<float>, std::vector<float>> Sphere::calculate_corners() const  {
    std::vector<float> min_corner = { static_cast<float>(_position.x) - static_cast<float>(_radius),
                                      static_cast<float>(_position.y) - static_cast<float>(_radius),
                                      static_cast<float>(_position.z) - static_cast<float>(_radius) };

    std::vector<float> max_corner = { static_cast<float>(_position.x) + static_cast<float>(_radius),
                                      static_cast<float>(_position.y) + static_cast<float>(_radius),
                                      static_cast<float>(_position.z) + static_cast<float>(_radius) };

    return std::make_pair(min_corner, max_corner);
}


std::pair<std::vector<float>, std::vector<float>> Sphere::calculate_minimum_box(Cell& perturbed_cell) const  {
    auto [cell1_min_corner, cell1_max_corner] = calculate_corners();
    auto [cell2_min_corner, cell2_max_corner] = perturbed_cell.calculate_corners();

    std::vector<float> min_corner, max_corner;
    for (int i = 0; i < 3; ++i) {
        min_corner.push_back(std::min(cell1_min_corner[i], cell2_min_corner[i]));
        max_corner.push_back(std::max(cell1_max_corner[i], cell2_max_corner[i]));
    }
    return std::make_pair(min_corner, max_corner);
}

bool Sphere::check_if_cells_overlap(const std::vector<Sphere>& spheres) {
    std::vector<std::vector<float>> positions;
    std::vector<float> radii;

    for (const auto& cell : spheres) {
        positions.push_back({ cell._position.x, cell._position.y, cell._position.z });
        radii.push_back(cell._radius * 0.95);
    }

    std::vector<std::vector<float>> distances;
    for (const auto& position1 : positions) {
        std::vector<float> distance_row;
        for (const auto& position2 : positions) {
            float distance = 0.0f;
            for (int i = 0; i < 3; ++i) {
                distance += pow(position1[i] - position2[i], 2);
            }
            distance = sqrt(distance);
            distance_row.push_back(distance);
        }
        distances.push_back(distance_row);
    }

    std::vector<std::vector<float>> radii_sums;
    for (const auto& radius1 : radii) {
        std::vector<float> radii_row;
        for (const auto& radius2 : radii) {
            radii_row.push_back(radius1 + radius2);
        }
        radii_sums.push_back(radii_row);
    }

    bool overlap = false;
    for (std::size_t i = 0; i < spheres.size(); ++i) {
        for (std::size_t j = 0; j < spheres.size(); ++j) {
            if (i != j && distances[i][j] < radii_sums[i][j]) {
                overlap = true;
                break;
            }
        }
        if (overlap) {
            break;
        }
    }

    return overlap;
}